### PR TITLE
Fixes to reduce memory usage for CPU backend in async mode. 

### PR DESCRIPTION
**Pull request by pavanky**

### CODE REVIEW

File: src/backend/MemoryManager.cpp [hunk 1]
```diff
@@ -258,7 +257,7 @@ void MemoryManager::printInfo(const char *msg, const int device)
             unit = "MB";
         }
 
-        std::cout << " |  " << std::right << std::setw(14) << kv.first << " "
+        std::cout << "|  " << std::right << std::setw(14) << kv.first << " "
```
Comment (umar456) [inline src/backend/MemoryManager.cpp]: I know this is not part of this PR but you should look into changing this into printfs. `cout` should only be used in templates.

Comment (pavanky) [inline src/backend/MemoryManager.cpp]: I didn't do it!

File: src/backend/cpu/kernel/sift_nonfree.hpp [hunk 2]
```diff
@@ -969,6 +969,7 @@ unsigned sift_impl(Array<float>& x, Array<float>& y, Array<float>& score,
                    const bool compute_GLOH)
 {
     in.eval();
+    getQueue().sync();
```
Comment (umar456) [inline src/backend/cpu/kernel/sift_nonfree.hpp]: Why is this required here?

File: src/backend/cpu/memory.cpp [hunk 3]
```diff
@@ -112,13 +112,30 @@ void printMemInfo(const char *msg, const int device)
 template<typename T>
 T* memAlloc(const size_t &elements)
 {
-    return (T *)getMemoryManager().alloc(elements * sizeof(T), false);
+    T *ptr = nullptr;
+
+    try {
```
Comment (umar456) [inline src/backend/cpu/memory.cpp]: Can't we return a null pointer in case of a failure here?.

File: src/backend/cpu/queue.hpp [hunk 4]
```diff
@@ -48,32 +49,45 @@ typedef async_queue queue_impl;
...
+    void enqueue(const F func, Args... args)
+    {
+        count++;
+        if(sync_calls) { func( args... ); }
+        else           { aQueue.enqueue( func, args... ); }
 #ifndef NDEBUG
-    sync();
+        sync();
+#else
+        if (checkMemoryLimit() || count >= 25) {
```
Comment (umar456) [inline src/backend/cpu/queue.hpp]: we should probably figure out a way to sync to a certain number of elements in the queue. Going all they way down to zero seems sub optimal

File: src/backend/cpu/scan.cpp [hunk 5]
```diff
@@ -27,8 +27,7 @@ template<af_op_t op, typename Ti, typename To>
 Array<To> scan(const Array<Ti>& in, const int dim)
 {
     dim4 dims     = in.dims();
-    Array<To> out = createValueArray<To>(dims, 0);
-    out.eval();
+    Array<To> out = createEmptyArray<To>(dims);
```
Comment (umar456) [inline src/backend/cpu/scan.cpp]: are you sure this does not expect 0 values.

**Comment by umar456** [comment]
couple of questions. Other then that it looks good.

File: src/backend/cpu/kernel/sift_nonfree.hpp (hunk 2 above)
Comment (pavanky) [inline src/backend/cpu/kernel/sift_nonfree.hpp]: The rest of the algorithm is synchronous. This function is not being called in a queue.

File: src/backend/cpu/memory.cpp (hunk 3 above)
Comment (pavanky) [inline src/backend/cpu/memory.cpp]: We return nullptr for 0 elements elsewhere. It'll get confusing.

File: src/backend/cpu/queue.hpp (hunk 4 above)
Comment (pavanky) [inline src/backend/cpu/queue.hpp]: This is a work around for now. We should probably remove it entirely by having a proper solution in the future.

File: src/backend/cpu/scan.cpp (hunk 5 above)
Comment (pavanky) [inline src/backend/cpu/scan.cpp]: I checked. I also ran the tests.

**Comment by pavanky** [comment]
Don't merge this yet. I am trying to see if I can fix this properly without using the work around.
